fontforge/autotrace.c

```diff
@@ -573,259 +473,231 @@ return;
+/**
+ * Performs an auto trace for a given character.
+ * 
+ * @param [in] sc The character to autotrace.
+ * @param [in] layer The layer to place the trace on.
+ * @param [in] ask Whether or not to ask for extra autotrace parameters.
+ * @return true if the autotrace was probably started.
+ */
+int SCAutoTrace(SplineChar *sc,int layer, int ask) {
     char **args;
 
-    if ( sc->layers[ly_back].images==NULL ) {
-	ff_post_error(_("Nothing to trace"),_("Nothing to trace"));
-return;
-    } else if ( FindAutoTraceName()==NULL ) {
-	ff_post_error(_("Can't find autotrace"),_("Can't find autotrace program (set AUTOTRACE environment variable) or download from:\n  http://sf.net/projects/autotrace/"));
-return;
+    if (sc->layers[ly_back].images == NULL) {
+        ff_post_error(_("Nothing to trace"),_("Nothing to trace"));
+        return false;
+    } else if (FindAutoTrace(NULL, 0) < 0) {
+        ff_post_error(_("Can't find autotrace"),_("Can't find autotrace program (set AUTOTRACE environment variable) or download from:\n  http://sf.net/projects/autotrace/"));
+        return false;
     }
 
-    args = AutoTraceArgs(ask);
-    if ( args==(char **) -1 )
-return;
+    args = QueryAutoTraceArgs(ask);
+    if (args==(char **) -1)
+        return false;
     _SCAutoTrace(sc, layer, args);
+    return true;
 }
 
-char *ProgramExists(const char *prog,char *buffer) {
-    char *path, *pt;
-
-    if (( path = getenv("PATH"))==NULL )
-return( NULL );
-
-    while ( 1 ) {
-	pt = strchr(path,':');
-	if ( pt==NULL ) pt = path+strlen(path);
-	if ( pt-path<1000 ) {
-	    strncpy(buffer,path,pt-path);
-	    buffer[pt-path] = '\0';
-	    if ( pt!=path && buffer[pt-path-1]!='/' )
-		strcat(buffer,"/");
-	    strcat(buffer,prog);
-	    /* Under cygwin, applying access to "potrace" will find "potrace.exe" */
-	    /*  no need for special check to add ".exe" */
-	    if ( access(buffer,X_OK)!=-1 ) {
-return( buffer );
-	    }
-	}
-	if ( *pt=='\0' )
-    break;
-	path = pt+1;
-    }
-return( NULL );
-}
-
-const char *FindAutoTraceName(void) {
-    static int searched=0;
-    static int waspotraceprefered;
-    static const char *name = NULL;
-    char buffer[1025];
 
-    if ( searched && waspotraceprefered==preferpotrace )
-return( name );
-
-    searched = true;
-    waspotraceprefered = preferpotrace;
-    if ( preferpotrace ) {
-	if (( name = getenv("POTRACE"))!=NULL )
-return( name );
+/**
+ * Finds the autotrace/potrace executable.
+ * 
+ * @param [out] path The buffer to store the location, or NULL if not required.
+ * @param [in] bufsiz The size of the output buffer.
+ * @return nonzero if potrace was found, zero if autotrace was found, negative
+ *         if none was found. Note previously, FontForge determined itself if
+ *         the resulting executable was potrace or autotrace, irrespective of
+ *         setting AUTOTRACE or POTRACE. (i.e. AUTOTRACE=potrace was possible).
+ */
+int FindAutoTrace(char *path, size_t bufsiz) {
```

Comment:
This seems a backward way of getting a string. Could we just return a dynamically allocated string or a null pointer?